### PR TITLE
Add support for PHP 8.4, Upgrade PHPUnit to 10.x, remove Prophecy dependency and improve unit tests

### PR DESCRIPTION
**Pull request by fezfez**

### CODE REVIEW

**Review by gsteel — CHANGES_REQUESTED** [review]
Thank you @fezfez :)

The code changes all look 👌 but can you please add to PHPUnit config `failOnDeprecation` along with `failOn…` `PhpUnitDeprecation|Error|Warning|Notice` and re-push - there are some deprecations in the tests that are currently going unnoticed and they may or may not be problematic

**Comment by fezfez** [comment]
@gsteel done expect failOnPhpunitDeprecation because of

```
2 tests triggered 2 PHPUnit deprecations:

1) LaminasTest\Mvc\DispatchListenerTest::testWillNotDispatchWhenAnMvcEventResultIsAlreadySet
Data Provider method LaminasTest\Mvc\DispatchListenerTest::alreadySetMvcEventResultProvider() is not static

/var/www/test/DispatchListenerTest.php:95

2) LaminasTest\Mvc\Service\ViewHelperManagerFactoryTest::testBasePathHelperFactoryCanBeInvokedViaShortNameOrFullClassName
Data Provider method LaminasTest\Mvc\Service\ViewHelperManagerFactoryTest::basePathConfiguration() is not static

/var/www/test/Service/ViewHelperManagerFactoryTest.php:134
```

those two static method create a mock with `$this->createMock(....)` and cant be marked as static, it will take some effort to fix this

**Comment by fezfez** [comment]
@gsteel fail on deprecation also fail on php 8.4 lowest job... Should a disable it too for now?

**Comment by gsteel** [comment]
Fixing the non-static static data providers will have to be done at some point - can you move the mocks to the test and provide any required data for the mock in the provider?

It's also helpful to set `displayDetailsOn{Deprecation|PhpUnitDepreaction|Notice|Warning}` in phpunit config so that we can see in the logs what failed.

**Comment by fezfez** [comment]
@gsteel this time i think it's good :)

**Review by gsteel — CHANGES_REQUESTED** [review]
Sorry @fezfez - I missed something, but other than that this is good to go. Thank you :)

File: src/ResponseSender/AbstractResponseSender.php [hunk 1]
```diff
@@ -19,16 +19,20 @@ public function sendHeaders(SendResponseEvent $event)
 
         $response = $event->getResponse();
 
-        foreach ($response->getHeaders() as $header) {
-            if ($header instanceof MultipleHeaderInterface) {
-                header($header->toString(), false);
-                continue;
+        $headers = $response->getHeaders();
+
+        if (is_array($headers) || is_iterable($headers)) {
```
Comment (gsteel) [inline src/ResponseSender/AbstractResponseSender.php]: I didn't notice this on my previous review, but arrays _are_ iterable, so just `is_iterable` suffices here: https://3v4l.org/JqP4X

**Review by gsteel — CHANGES_REQUESTED** [review]
🤦‍♂️ Forgot to add the review comment

File: src/ResponseSender/AbstractResponseSender.php (hunk 1 above)
Comment (fezfez) [reply src/ResponseSender/AbstractResponseSender.php]: good catch, done

**Review by gsteel — APPROVED** [review]
Thanks again @fezfez

**Pull request merged**